promql/engine.go

```diff
@@ -2141,3 +2253,163 @@ func unwrapParenExpr(e *parser.Expr) {
 		}
 	}
 }
+
+func unwrapStepInvariantExpr(e parser.Expr) parser.Expr {
+	if p, ok := e.(*parser.StepInvariantExpr); ok {
+		return p.Expr
+	}
+	return e
+}
+
+// WrapWithStepInvariantExpr wraps all possible parts of the given
+// expression with StepInvariantExpr wherever valid.
+func WrapWithStepInvariantExpr(expr parser.Expr) parser.Expr {
+	isStepInvariant := wrapWithStepInvariantExprHelper(expr)
+	if isStepInvariant {
+		return newStepInvariantExpr(expr)
+	}
+	return expr
+}
+
+// wrapWithStepInvariantExprHelper wraps the child nodes of the expression
+// with a StepInvariantExpr wherever valid. The returned boolean is true if the
+// passed expression qualifies to be wrapped by StepInvariantExpr.
+func wrapWithStepInvariantExprHelper(expr parser.Expr) bool {
+	switch n := expr.(type) {
+	case *parser.VectorSelector:
+		return n.Timestamp != nil
+
+	case *parser.AggregateExpr:
+		return wrapWithStepInvariantExprHelper(n.Expr)
+
+	case *parser.BinaryExpr:
+		isInvariant1, isInvariant2 := wrapWithStepInvariantExprHelper(n.LHS), wrapWithStepInvariantExprHelper(n.RHS)
+		if isInvariant1 && isInvariant2 {
+			return true
+		}
+
+		if isInvariant1 {
+			n.LHS = newStepInvariantExpr(n.LHS)
+		}
+		if isInvariant2 {
+			n.RHS = newStepInvariantExpr(n.RHS)
+		}
+
+		return false
+
+	case *parser.Call:
+		isStepInvariant := IsFunctionStepInvariant(n)
+		isStepInvariantSlice := make([]bool, len(n.Args))
+		for i := range n.Args {
+			isStepInvariantSlice[i] = wrapWithStepInvariantExprHelper(n.Args[i])
+			isStepInvariant = isStepInvariant && isStepInvariantSlice[i]
+		}
+
+		if n.Func.Name == "timestamp" {
+			// If timestamp(..) has something other than vector selector inside,
```

Comment:
This is wrong, timestamp is always step invariant - if you give it the same input data it always produces the same output, independent of the evaluation time. This code shouldn't be here.

Reply:
If the test case for `timestamp(timestamp(...))` is right, then we need this special casing to not wrap the outer `timestamp()`. Or do you mean the timestamp of the vector selector is propagated through the nested timestamp()? For that matter, any expression that is inside the timestamp() which is not an instant vector selector, wont that get a timestamp same as eval time and hence should not wrap the timestamp() if it has non instant vector selector inside?

Reply:
While the timestamp() itself is step invariant, the expression is not, hence this special casing

Reply:
> Or do you mean the timestamp of the vector selector is propagated through the nested timestamp()?

I mean that if you think you need this, something else is broken somewhere. We should find what that is and fix that, rather than hacking around it.

Reply:
I think this is the place where it should be fixed, because, it is the wrapping and duplication of result that is not allowing nested timestamp() and similar to function in a way it should. We cannot really do that during duplication.